File: 15. 3Sum/three_sum.py
from typing import List

def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        for i,a in enumerate(nums):
            if i>0 and a == nums[i-1]:
                continue
            j,k = i+1, len(nums)-1
            while k>j:
                sum = a + nums[j] + nums[k]
                if sum > 0:
                    k-=1
                elif sum < 0:
                    j+=1
                else:
                    res.append([a,nums[j], nums[k]])
                    j+=1
                    while nums[j] == nums[j-1] and j<k:
                        j+=1
        return res

# 12/Jan/2022
# Runtime: 788 ms, faster than 83.36% of Python3 online submissions for 3Sum.
# Memory Usage: 17.4 MB, less than 87.05% of Python3 online submissions for 3Sum.

File: 0075 Must Do Leetcode/0015. 3Sum/ThreeSum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
    nums.sort((a,b) => a-b);
    let res = [];
    for (var i=0; i < nums.length; i++) {
        if (i > 0 && nums[i] == nums[i - 1])
            continue;
        let [j, k] = [i+1, nums.length - 1];
        while (k>j) {
            let sum = nums[i] + nums[j] + nums[k];

            if (sum > 0)
                k--;
            else if (sum < 0)
                j++;
            else {
                res.push([nums[i], nums[j], nums[k]])
                j++;
                while (nums[j] == nums[j-1] && j < k) {
                    j++;
                }
            }
        }
    }

    return res;
};

for (var r of threeSum([-1, 0, 1, 2, -1, -4])) {
    console.log(r);
}
console.log("=========")
for (var r of threeSum([-1,0,1,2,-1,-4,-2,-3,3,0,4])) {
    console.log(r);
}
console.log("=========")
for (var r of threeSum([])) {
    console.log(r);
}
console.log("=========")
for (var r of threeSum([-1,0,1,0])) {
    console.log(r);
}
console.log("=========")

// 12/Jan/2022
// Runtime: 331 ms, faster than 28.74% of JavaScript online submissions for 3Sum.
// Memory Usage: 50.3 MB, less than 29.51% of JavaScript online submissions for 3Sum.